src/coreclr/src/pal/src/exception/machexception.cpp

```diff
@@ -1429,6 +1429,19 @@ ActivationHandler(CONTEXT* context)
     DebugBreak();
 }
 
+#if defined(TARGET_ARM64)
+extern "C"
+void
+RtlRestoreContext(CONTEXT* context, PEXCEPTION_RECORD ExceptionRecord)
```

Comment:
```asm
    str x17, [sp]
    ldp x16, x17, [x16, CONTEXT_X16]
    ldr pc, [sp]
// Unreachable ...  pc changed...
LOCAL_LABEL(No_Restore_CONTEXT_CONTROL):
    ldp x16, x17, [x16, CONTEXT_X16]
    ret
```

Reply:
It might be better to alter the JIT to avoid using X17 & (maybe X16 if needed).

Reply:
I don't understand why my code doesn't work -  it is intentional that we don't fall through to the ret. We were not falling through there before either. The `No_Restore_CONTEXT_CONTROL` is there for the cases when restoring of control registers was not requested, so we return back to the caller.

Reply:
Actually, I have a slightly better version. The one I've posted above uses redzone of the target frame, which might be a problem in case JIT started to use redzone in the future. Here is a version that uses the redzone of the frame of the RtlRestoreContext, which is always safe:
```
    ldr x17, [x16, CONTEXT_Sp]
    str x17, [sp]
    ldr x17, [x16, CONTEXT_Pc]
    str x17, [sp, -4]
    ldp x16, x17, [x16, CONTEXT_X16]
    ldp sp, pc, [sp, -4]

```

Reply:
> I don't understand why my code doesn't work

Sorry, I didn't read/write carefully enough.  There is no `ldr pc...` or `ldp pc ...` because the `pc` is not a register in `aarch64`. `pc` can only be affected by a branch literal, branch register (including ret), and ERET (not available at EL0).

So if we JITTED code for the `pc` return with a target that was in range, I think it might be possible, but that requires kernel calls to toggle JIT write, code management ... Feels too complicated.

The `RtlCaptureContext` currently has similar issues of not capturing X16/X17.  I think that is probably fixable though.  Especially since RtlCapture could have a frame.

This problem was discovered during testing activation.  The original draft-fix  changed the `ActivationHandler` to use `MachSetThreadContext(context);` for `arm64`.  That might work...  Activation is somewhat different because we stop the thread and redirect it from another thread. So we need a perfect restore.

I suspect the other uses of RtlCapture/RtlRestore are more forgiving (because they are call sites) ...  I just hadn't done a full audit...

Reply:
I have looked at that before and I also believe that the only place where we need the full fidelity capture is the ActivationHandler.

Reply:
As for the `ldp sp, pc, [sp, -4]`, it is funny that the assembler happily compiled that instruction. Sigh.